### DISCUSSION TITLE
Import extra WASI Functions

### DISCUSSION THREAD

**Discussion by mandesero** [konsoletyper/teavm#968] [Q&A]
I'm trying to add WebAssembly networking support to a Java project using **TeaVM**. TeaVM does support importing WASI functions, as demonstrated in the [`Wasi.java`](https://github.com/konsoletyper/teavm/blob/master/core/src/main/java/org/teavm/backend/wasm/wasi/Wasi.java) class, which successfully imports functions like `clock_time_get` and `fd_read` using `@Import`.

However, when I try to use a similar mechanism in my custom `WasiSocket` class to import additional WASI networking functions, such as `sock_accept` and `sock_addr_local`, I encounter issues. Here’s the `WasiSocket` class:

```java
package org.teavm.backend.wasm.wasi;

import org.teavm.interop.Import;

public class WasiSocket {
    @Import(name = "sock_accept", module = "wasi_snapshot_preview1")
    public static native int sockAccept(int fd, int flags, int fdNew);

    @Import(name = "sock_addr_local", module = "wasi_snapshot_preview1")
    public static native int sockAddrLocal(int fd, WasiAddr addr);
    // other WASI networking functions...
}
```

When I try to run the WebAssembly module generated by TeaVM in **Wasmtime** or in a **browser environment**, I get the following error:

```
Failed to load the WASM module [TypeError: WebAssembly.instantiate(): Import #6 module="wasi_snapshot_preview1" error: module is not an object or function]
```

This suggests that my imports for WASI networking functions are not being recognized correctly, even though a similar approach works for other WASI functions in `Wasi.java`.

Here’s my setup:
- **TeaVM fork with changes**: [https://github.com/mandesero/teavm](https://github.com/mandesero/teavm)
- **Example project**: [https://github.com/mandesero/teavm-wasm-examples](https://github.com/mandesero/teavm-wasm-examples)

How can I make these additional WASI networking imports work in TeaVM? Are there specific configurations, limitations, or adjustments required for importing extra WASI functions in TeaVM?

**Comment by konsoletyper** [comment]
First of all, in order to use `@Import` annotation, you don't necessary need to modify TeaVM. You can use vanilla TeaVM and declare native methods, annotated with `@Import` annotation, and this will work *exactly* as if with `Wasi` class.

> When I try to run the WebAssembly module generated by TeaVM in Wasmtime or in a browser environment, I get the following error:
>
> ```
> Failed to load the WASM module [TypeError: WebAssembly.instantiate(): Import #6 module="wasi_snapshot_preview1" error: module is not an object or function]
> ```
>
> This suggests that my imports for WASI networking functions are not being recognized correctly, even though a similar approach works for other WASI functions in Wasi.java.

Yes, looks like these imports are not recognized correctly *with your WebAssembly environment*, perhaps TeaVM has nothing to do with this (unless it's generating some broken wasm file, which, perhaps, is not the case). So you should further investigate why your environment does not export these functions.
